http://www.openwall.com/lists/musl/2018/01/06/3

Date: Sat,  6 Jan 2018 23:32:52 +0100
From: Hauke Mehrtens <hauke@...ke-m.de>
To: musl@...ts.openwall.com
Cc: felix.janda@...teo.de,
	Hauke Mehrtens <hauke@...ke-m.de>
Subject: [PATCH v2] Add additional uapi guards for Linux kernel header files

With Linux kernel 4.16 it will be possible to guard more parts of the
Linux header files from a libc. Make use of this in musl to guard all
the structures and other definitions from the Linux header files which
are also defined by the header files provided by musl. This will make
musl compile with the unmodified Linux kernel user space headers.

This extends the definitions done in commit 04983f227238 ("make
netinet/in.h suppress clashing definitions from kernel headers")

The needed patches were recently accepted into the netdev tree and will be integrated in Linux 4.16:
https://patchwork.ozlabs.org/patch/854342/
https://patchwork.ozlabs.org/patch/855293/
---
 include/net/if.h           | 7 +++++++
 include/netinet/if_ether.h | 1 +
 include/sys/xattr.h        | 2 ++
 3 files changed, 10 insertions(+)

diff --git a/include/net/if.h b/include/net/if.h
index 2f2fcc10..0ee48cd7 100644
--- a/include/net/if.h
+++ b/include/net/if.h
@@ -125,6 +125,13 @@ struct ifconf {
 #define ifc_req		ifc_ifcu.ifcu_req
 #define _IOT_ifconf _IOT(_IOTS(struct ifconf),1,0,0,0,0)
 
+#define __UAPI_DEF_IF_IFCONF					0
+#define __UAPI_DEF_IF_IFMAP					0
+#define __UAPI_DEF_IF_IFNAMSIZ					0
+#define __UAPI_DEF_IF_IFREQ					0
+#define __UAPI_DEF_IF_NET_DEVICE_FLAGS				0
+#define __UAPI_DEF_IF_NET_DEVICE_FLAGS_LOWER_UP_DORMANT_ECHO	0
+
 #endif
 
 #ifdef __cplusplus
diff --git a/include/netinet/if_ether.h b/include/netinet/if_ether.h
index d9a131aa..c2c6e944 100644
--- a/include/netinet/if_ether.h
+++ b/include/netinet/if_ether.h
@@ -133,5 +133,6 @@ do { \
 	(enaddr)[5] = ((uint8_t *)ipaddr)[3]; \
 } while(0)
 
+#define __UAPI_DEF_ETHHDR	0
 
 #endif
diff --git a/include/sys/xattr.h b/include/sys/xattr.h
index 6479fcc6..52e3dd89 100644
--- a/include/sys/xattr.h
+++ b/include/sys/xattr.h
@@ -24,6 +24,8 @@ int removexattr(const char *, const char *);
 int lremovexattr(const char *, const char *);
 int fremovexattr(int, const char *);
 
+#define __UAPI_DEF_XATTR	0
+
 #ifdef __cplusplus
 }
 #endif
-- 
2.11.0
